packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/mod.rs

```diff
@@ -148,6 +148,53 @@ impl<C: AsClient> EntityStore for PostgresStore<C> {
         Ok(identifier)
     }
 
+    async fn create_entities(
+        &mut self,
+        entities: impl IntoIterator<Item = (Option<EntityId>, Entity), IntoIter: Send> + Send,
+        entity_type_id: VersionedUri,
+        owned_by_id: AccountId,
+    ) -> Result<Vec<EntityId>, InsertionError> {
+        let transaction = PostgresStore::new(
+            self.as_mut_client()
+                .transaction()
+                .await
+                .into_report()
+                .change_context(InsertionError)?,
+        );
+
+        let (entity_ids, entities): (Vec<_>, Vec<_>) = entities
+            .into_iter()
+            .map(|(id, entity)| (id.unwrap_or_else(|| EntityId::new(Uuid::new_v4())), entity))
+            .unzip();
+
+        // TODO: match on and return the relevant error
+        //   https://app.asana.com/0/1200211978612931/1202574350052904/f
+        transaction
```

Comment:
I'm also concerned about the safety of the inner methods if they aren't wrapped in a transaction. I'm not confident right now that this is going to be safe, and I'm also not happy with the level of documentation given to the methods.

I think it'd be too easy for us to change logic, or implement new logic, which ends up with entity IDs in the store that aren't associated with entities, etc.

That's based on a very quick think though, is my worry unfounded? 
1. As in do we already ensure referential integrity at the Postgres level and this will be checked as part of the transaction?
2. Would calling `insert_entity_ids` by itself succeed?

Reply:
cc @thehabbos007, and cc @benwerner01 (I appreciate there's a lot of somewhat complicated Rust here but we should perhaps discuss offline how we actually want to handle bulk inserts at the conceptual level, as in talking about transactions, Postgres integrity guarantees, etc.)

Reply:
Calling these batch operations outside transactions might be somewhat of an issue, as you say we could have dangling entity IDs if entity insertion fails and we aren't using a transaction around these. Without having type-level safety about transactions, this would creep up for any sort of bulk operation we want to atomically apply (perhaps link re-ordering would also fall under this category).

The main issue to me seems to be that of dangling IDs on unsuccessful insertions. The ID insertion would likely never fail, whereas the entity insertion is driven by user input, and therefore we could very well have dangling IDs outside a transaction - which wouldn't leave the database in an invalid state, just in a undesired state.

I think for things like bulk inserting entities with links, we could end up in invalid state e.g. link pointing to an entity that has no entities entry, meaning the link would result in an invalid join.

Reply:
These methods can only be called inside of `PostgresStore<Transaction>`, otherwise the `copy_in` method is not available. So without creating a transaction, it wouldn't compile.

Reply:
> So without creating a transaction, it wouldn't compile.

This is good to know 👍 , although having to rely on that being the marker seems a little obscure vs some explicit expression of being inside a transaction